test/src/data/article/ny_times_paginated_rest_repository_test.dart

```diff
@@ -0,0 +1,96 @@
+import 'package:allthenews/src/data/article/ny_times_paginated_rest_repository.dart';
+import 'package:allthenews/src/data/communication/api/http_client.dart';
+import 'package:allthenews/src/domain/common/page.dart';
+import 'package:allthenews/src/domain/nytimes/ny_times_cached_repository.dart';
+import 'package:allthenews/src/domain/nytimes/ny_times_repository.dart';
+import 'package:allthenews/src/domain/settings/settings_repository.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:mockito/mockito.dart';
+
+class MockHttpClient extends Mock implements HttpClient {}
+
+class MockSettingsRepository extends Mock implements SettingsRepository {}
+
+class MockNyTimesRepository extends Mock implements NYTimesRepository {}
+
+class MockNyTimesCachedRepository extends Mock implements NyTimesCachedRepository {}
+
+void main() {
+  NyTimesPaginatedRestRepository nyTimesPaginatedRestRepository;
+  MockNyTimesRepository mockNyTimesRepository;
+  MockNyTimesCachedRepository mockNyTimesCachedRepository;
+  MockSettingsRepository mockSettingsRepository;
+
+  setUp(() {
+    mockNyTimesRepository = MockNyTimesRepository();
+    mockNyTimesCachedRepository = MockNyTimesCachedRepository();
+    mockSettingsRepository = MockSettingsRepository();
+    nyTimesPaginatedRestRepository = NyTimesPaginatedRestRepository(
+      mockNyTimesRepository,
+      mockNyTimesCachedRepository,
+      mockSettingsRepository,
+    );
+  });
+
+  group('latest news', () {
```

Comment:
te testy moim zdaniem mało wnoszą wartości, sprawdzamy czy wykona się `if` jeżeli podamy `1` jako parametr wywołania. Bardziej przydatne byłoby przetestowanie czy po pobraniu z api danych wyciągniemy takie same dane z bazy. Może można by było postawić wirtualną bazę i rzeczywiście tam wrzucać te dane? https://moor.simonbinder.eu/docs/testing/

Reply:
a tutaj się nie zgodzę, sprawdzanie czy wywoła się if to duże spłycenie. Generalnie sprawdzanie polega na tym czy jak wołasz o pierwszą stronę to czy pierw pobiera z rest repo, potem zapisuje do cache i dopiero wyciąga z cache, natomiast jak bije po kolejną stronę, to już żeby nie bił do rest repo ani nic nie cachował tylko zwrócił dane z cache, tak, żeby np. nie zaszła jakaś niespójność danych. To, że w kodzie sprowadza się do ifa nie ma znaczenia, ważne, że działa zgodnie z logiką i tę logikę testuje. Co do wirtualnej bazy to ogólnie tutaj testowane jest repo, więc najmniejszą jednostką jest to repo i jego zależności zostały zamockowane, tak, żeby przetestować repository. Testowanie w oparciu o wirtualną bazę testowałoby repository + articleDao + sprawdzanie czy to co wrzucisz do bazy to czy to zwróci to już testowałoby samą implementację bazy. Co do wirtualnej bazy to próbowałem w oparciu o to przetestować ArticleDao, ale bez skutku, bo generalnie stwierdziłem, że nie ma sensu z tego powodu: https://github.com/simolus3/moor/issues/78#issuecomment-547065933 Jeżeli trzeba dopinać sqlite, żeby wykonać testy jednostkowe to znaczy, że chyba coś źle zaprojektowali

Reply:
> Generalnie sprawdzanie polega na tym czy jak wołasz o pierwszą stronę to czy pierw pobiera z rest repo, potem zapisuje do cache i dopiero wyciąga z cache, natomiast jak bije po kolejną stronę, to już żeby nie bił do rest repo ani nic nie cachował tylko zwrócił dane z cache, tak, żeby np. nie zaszła jakaś niespójność danych.

Zgadzam się, że dobrze jest to sprawdzić, chociażby dlatego, że sposób w jaki to wyciągamy wewnątrz metody może się zmienić, teraz jest to `if`, ale w przyszłości może być to coś bardziej złożonego. Przy obecnej implementacji testów nie sprawdzimy logiki metody, ale to czy kod się zmienił. Test, który wrzucał by na wejściu do dao dane, a potem wykonywał metodę i sprawdzał jakie dane dostał w zależności od przekazanej strony byłby bardziej odporny na zmiany

Reply:
Ok, pomyślałem i przyznaję rację, że wówczas testy będą mniej kruche :) Niemniej jednak wciąż uważam, że logika została przetestowana, bo jeżeli ten test sprawdza tylko czy kod się zmienił to wszystkie verify, verifyNever, verifyInOrder w Mockito są zbędne 🤔 Zaraz wrzucę poprawiony test, tylko nie wrzucam danych do dao, bo wtedy testowane by było niepotrzebnie NyTimesCachedRepository :)